### PR TITLE
#3787 handle different sources of migrations files

### PR DESCRIPTION
**Pull request by dcdh**
In case migration files are presents on an external jar, the external
jar must be unzipped to avoid FileSystemNotFoundException

### CODE REVIEW

**Comment by geoand** [comment]
CI is failing with the PR

**Comment by dcdh** [comment]
> CI is failing with the PR

Yes I know, I've push a commit however the PR refuse to be updated :S I don't know why ...

**Comment by geoand** [comment]
> > CI is failing with the PR
> 
> Yes I know, I've forced push a commit however the PR refuse to be updated :S I don't know why ...

Are you sure the change got pushed? Github UI is not showing it. Moreover github [status](https://www.githubstatus.com/) seems to report everything is working on their end.

**Comment by dcdh** [comment]
> > > CI is failing with the PR
> > 
> > 
> > Yes I know, I've forced push a commit however the PR refuse to be updated :S I don't know why ...
> 
> Are you sure the change got pushed? Github UI is not showing it. Moreover github [status](https://www.githubstatus.com/) seems to report everything is working on their end.

I mix with my branch... I will make right change now.

**Comment by geoand** [comment]
Thanks for the contribution!

This looks good to me, but I'll let someone who has knows more about this piece of code review.

**Comment by dcdh** [comment]
> Thanks for the contribution!
> 
> This looks good to me, but I'll let someone who has knows more about this piece of code review.

Unfortunately I've got an issue locally. I'll keep you in touch.

**Comment by dcdh** [comment]
I have notice that the protocol is a good way to know if migrations files are presents in an external jar or directly in the resources of the main application.
However I don't find a way to properly add an automatic test for the case where migrations files are presents in an external jar.

**Comment by dcdh** [comment]
CI seems to be down. Can someone start a rebuild ? Thanks by advance.

**Comment by geoand** [comment]
Done

**Comment by stuartwdouglas** [comment]
oops, I broke CI. You will need to rebase on the latest master to get a clean run

File: extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java [hunk 1]
```diff
@@ -35,6 +40,10 @@
 import io.quarkus.flyway.runtime.graal.QuarkusPathLocationScanner;
 
 class FlywayProcessor {
+
+    private static final String EXTERNAL_APPLICATION_MIGRATIONS_PROTOCOL = "jar";
+    private static final String INTERNAL_APPLICATION_MIGRATIONS_PROTOCOL = "file";
```
Comment (gsmet) [inline extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java]: I would name them `JAR_...` and `FILE_...`

File: extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java [hunk 2]
```diff
@@ -144,6 +163,21 @@ private void registerSubstrateResources(BuildProducer<SubstrateResourceBuildItem
         }
     }
 
+    private Set<String> getApplicationMigrationsFromPath(final String location, final URL path)
+            throws IOException, URISyntaxException {
+        return walk(Paths.get(path.toURI()))
+                .filter(Files::isRegularFile)
+                .map(it -> Paths.get(location, it.getFileName().toString()).toString())
```
Comment (gsmet) [inline extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java]: This was preexisting but I'm a bit puzzled by this line. I wonder what it's supposed to do.

**Review by gsmet — CHANGES_REQUESTED** [review]
Added a couple of comments/questions.

File: extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java (hunk 2 above)
Comment (dcdh) [reply extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java]: It construct a Set of all migration files compound of a location and each files presents in this location.
For example: **01_init.sql** and **02_update.sql** presents in the location **db/migration** will produce a Set of String containing : **db/migration/01_init.sql** and **db/migration/02_update.sql**.
map

**Comment by dcdh** [comment]
I've rewrote history to fix logger issue in my first commit.

**Comment by dcdh** [comment]
On this [project](https://github.com/dcdh/poc-quarkus-external-jar) I have fixed this issue by overloading the classpath : FlywayProcessor has been created on the `helloworld-app` project.

At startup the log shows:

> [INFO] [io.quarkus.flyway.FlywayProcessor] Adding application migrations in path '/home/damien/IdeaProjects/pocquarkusexternaljar/helloworld-app/target/classes/db/migration' using protocol 'file'
> [INFO] [io.quarkus.flyway.FlywayProcessor] Adding application migrations in path 'file:/home/damien/.m2/repository/com/damdamdeo/helloworld-service/1.0-SNAPSHOT/helloworld-service-1.0-SNAPSHOT.jar!/db/migration' using protocol 'jar'

and the application starts and data are fully loaded by flyway ^^

File: extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java (hunk 2 above)
Comment (gsmet) [reply extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java]: Ah OK, I see, they are only used for native resources registration so you need relative paths.

**Comment by gsmet** [comment]
@dcdh could you squash at least the first 2 commits? Thanks.

**Comment by dcdh** [comment]
@gsmet I have squashed the first 2 commits.

**Review by gsmet — APPROVED** [review]

**Pull request merged**